pkg/resource/models/service_resource.go

```diff
@@ -0,0 +1,125 @@
+		"protocol": e.Protocol,
+		"host":     e.Host,
+		"port":     e.Port,
+	})
+}
+
+type IpAccessList struct {
+	Source      types.String `tfsdk:"source"`
+	Description types.String `tfsdk:"description"`
+}
+
+func (i IpAccessList) ObjectType() types.ObjectType {
+	return types.ObjectType{
+		AttrTypes: map[string]attr.Type{
+			"source":      types.StringType,
+			"description": types.StringType,
+		},
+	}
+}
+
+func (i IpAccessList) ObjectValue() basetypes.ObjectValue {
+	return types.ObjectValueMust(i.ObjectType().AttrTypes, map[string]attr.Value{
+		"source":      i.Source,
+		"description": i.Description,
+	})
+}
+
+type PrivateEndpointConfig struct {
+	EndpointServiceID  types.String `tfsdk:"endpoint_service_id"`
+	PrivateDNSHostname types.String `tfsdk:"private_dns_hostname"`
+}
+
+func (p PrivateEndpointConfig) ObjectType() types.ObjectType {
+	return types.ObjectType{
+		AttrTypes: map[string]attr.Type{
+			"endpoint_service_id":  types.StringType,
+			"private_dns_hostname": types.StringType,
+		},
+	}
+}
+
+func (p PrivateEndpointConfig) ObjectValue() basetypes.ObjectValue {
+	return types.ObjectValueMust(p.ObjectType().AttrTypes, map[string]attr.Value{
+		"endpoint_service_id":  p.EndpointServiceID,
+		"private_dns_hostname": p.PrivateDNSHostname,
+	})
+}
+
+type ServiceResourceModel struct {
+	ID                              types.String `tfsdk:"id"`
+	Name                            types.String `tfsdk:"name"`
+	Password                        types.String `tfsdk:"password"`
+	PasswordHash                    types.String `tfsdk:"password_hash"`
+	DoubleSha1PasswordHash          types.String `tfsdk:"double_sha1_password_hash"`
+	Endpoints                       types.List   `tfsdk:"endpoints"`
+	CloudProvider                   types.String `tfsdk:"cloud_provider"`
+	Region                          types.String `tfsdk:"region"`
+	Tier                            types.String `tfsdk:"tier"`
+	IdleScaling                     types.Bool   `tfsdk:"idle_scaling"`
+	IpAccessList                    types.List   `tfsdk:"ip_access"`
+	MinTotalMemoryGb                types.Int64  `tfsdk:"min_total_memory_gb"`
+	MaxTotalMemoryGb                types.Int64  `tfsdk:"max_total_memory_gb"`
+	NumReplicas                     types.Int64  `tfsdk:"num_replicas"`
+	IdleTimeoutMinutes              types.Int64  `tfsdk:"idle_timeout_minutes"`
+	IAMRole                         types.String `tfsdk:"iam_role"`
+	LastUpdated                     types.String `tfsdk:"last_updated"`
+	PrivateEndpointConfig           types.Object `tfsdk:"private_endpoint_config"`
+	PrivateEndpointIds              types.List   `tfsdk:"private_endpoint_ids"`
+	EncryptionKey                   types.String `tfsdk:"encryption_key"`
+	EncryptionAssumedRoleIdentifier types.String `tfsdk:"encryption_assumed_role_identifier"`
+}
+
+func (m *ServiceResourceModel) Equals(b ServiceResourceModel) bool {
+	if !m.ID.Equal(b.ID) ||
```

Comment:
An idea for next step: these Equal functions can be generated with something like https://github.com/awalterschulze/goderive/tree/main/example/plugin/equal

Reply:
nice one